### PR TITLE
[clangd] Support rename of Objective-C method names

### PR DESCRIPTION
**Pull request by ahoppen**
This integrates the rename support for Objective-C selectors that we had in libclang into `clangd`. 

While doing this, it unifies `SymbolName` and `OldSymbolName`. `OldSymbolName` was used by `Refactor`, the refactoring library that only exists in Apple’s fork, while `SymbolName` was used by `Refactoring`, the upstream refactoring library. Unify them and add mulit-piece symbol support ot the upstream `SymbolName`.

### Upstreaming plan
- [ ] After this PR is merged, I’ll merge the same to `next`
- [ ] Once we have the PR in `next`, I’ll create a PR to LLVM that upstreams the changes. I cannot start with the upstream LLVM PR because we need to make sure that this fork continues to build with the changes, ie. libclang’s rename (which hasn’t been upstreamed) should be able to use the same core functionality

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

File: clang/lib/Tooling/Refactor/RenameIndexedFile.cpp [hunk 1]
```diff
@@ -256,28 +257,36 @@ SourceLocation StringLiteralTextualParser::handleToken(const Token &RawTok,
...
   if (Literal.hadError)
     return SourceLocation();
-  return Literal.GetString() == Name[0]
+  return Literal.GetString() == Name.getNamePieces()[0]
              ? RawTok.getLocation().getLocWithOffset(
                    Literal.getOffsetOfStringByte(RawTok, 0))
              : SourceLocation();
 }
 
+static bool containsEmptyPiece(const SymbolName &Name) {
```
Comment (bnbarham) [inline clang/lib/Tooling/Refactor/RenameIndexedFile.cpp:268]: Any reason not to add this to `SymbolName`?

File: clang/lib/Tooling/Refactoring/SymbolName.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  for (const auto &Piece : NamePieces)
+    this->NamePieces.push_back(Piece.str());
+}
+
+void SymbolName::print(raw_ostream &OS) const {
+  for (size_t I = 0, E = NamePieces.size(); I != E; ++I) {
+    if (I != 0)
+      OS << ':';
+    OS << NamePieces[I];
+  }
```
Comment (bnbarham) [inline clang/lib/Tooling/Refactoring/SymbolName.cpp]:
```
llvm::interleave(NamePieces, OS, ":");
```

File: clang/include/clang/Tooling/Refactoring/Rename/SymbolName.h [hunk 3]
```diff
@@ -27,19 +30,22 @@ namespace tooling {
 /// //       ^~ string 0 ~~~~~         ^~ string 1 ~~~~~
 /// \endcode
 class SymbolName {
```
Comment (bnbarham) [inline clang/include/clang/Tooling/Refactoring/Rename/SymbolName.h:33]: Could split this up into a separate commit to make upstreaming a little easier.

File: clang-tools-extra/clangd/refactor/Rename.cpp [hunk 4]
```diff
@@ -517,25 +521,28 @@ static bool mayBeValidIdentifier(llvm::StringRef Ident) {
...
-  else {
-    // Name conflict detection.
-    // Function conflicts are subtle (overloading), so ignore them.
-    if (RenameDecl.getKind() != Decl::Function &&
-        RenameDecl.getKind() != Decl::CXXMethod) {
-      if (auto *Conflict = lookupSiblingWithName(ASTCtx, RenameDecl, NewName))
-        Result = InvalidName{
-            InvalidName::Conflict,
-            Conflict->getLocation().printToString(ASTCtx.getSourceManager())};
+  if (auto Identifier = NewName.getSinglePiece()) {
```
Comment (bnbarham) [inline clang-tools-extra/clangd/refactor/Rename.cpp]: Personally I'd do the following to avoid the nesting level:
```
auto Identifier = NewName.getSinglePiece();
if (!Identifier)
 return Result;
```

File: clang-tools-extra/clangd/refactor/Rename.cpp [hunk 5]
```diff
@@ -569,9 +576,28 @@ renameWithinFile(ParsedAST &AST, const NamedDecl &RenameDecl,
...
     if (!isInsideMainFile(RenameLoc, SM))
       continue;
-    if (auto Err = FilteredChanges.add(tooling::Replacement(
-            SM, CharSourceRange::getTokenRange(RenameLoc), NewName)))
-      return std::move(Err);
+    if (std::optional<std::string> Identifier = NewName.getSinglePiece()) {
+      tooling::Replacement NewReplacement(
+          SM, CharSourceRange::getTokenRange(RenameLoc), *Identifier);
+      if (auto Err = FilteredChanges.add(NewReplacement))
+        return std::move(Err);
```
Comment (bnbarham) [inline clang-tools-extra/clangd/refactor/Rename.cpp:585]: Can we just early return `FilteredChanges` here and unnest the else?

File: clang-tools-extra/clangd/refactor/Rename.cpp [hunk 6]
```diff
@@ -569,9 +576,28 @@ renameWithinFile(ParsedAST &AST, const NamedDecl &RenameDecl,
...
+    } else {
+      llvm::Expected<SmallVector<SourceLocation>> PieceLocations =
+          findObjCSymbolSelectorPieces(
+              AST.getTokens().expandedTokens(), SM, RenameLoc, OldName,
+              tooling::ObjCSymbolSelectorKind::Unknown);
+      if (!PieceLocations) {
+        return PieceLocations.takeError();
+      }
+      assert(PieceLocations->size() == NewName.getNamePieces().size());
```
Comment (bnbarham) [inline clang-tools-extra/clangd/refactor/Rename.cpp]: `zip_equal` already has this assertion

File: clang-tools-extra/clangd/unittests/RenameTests.cpp [hunk 7]
```diff
@@ -926,12 +975,12 @@ TEST(RenameTest, Renameable) {
...
 
-      {R"cpp(// disallow rename on non-normal identifiers.
+      {R"cpp(
          @interface Foo {}
-         -(int) fo^o:(int)x; // Token is an identifier, but declaration name isn't a simple identifier.
+         -(int) [[fo^o]]:(int)x;
          @end
        )cpp",
-       "not a supported kind", HeaderFile},
+       nullptr, HeaderFile},
```
Comment (bnbarham) [inline clang-tools-extra/clangd/unittests/RenameTests.cpp:1085]: Why's the error lost?

File: clang/include/clang/Tooling/Refactoring/Rename/RenamingAction.h [hunk 8]
```diff
@@ -116,6 +117,26 @@ class QualifiedRenamingAction {
...
+  MethodDecl,
+
+  /// It is unknown if the renamed location is a method call or declaration.
+  ///
+  /// The selector kind is being used to improve error recovery, passing unknown
+  /// does not lead to correctness issues.
+  Unknown
+};
+
+llvm::Expected<SmallVector<SourceLocation>> findObjCSymbolSelectorPieces(
```
Comment (bnbarham) [inline clang/include/clang/Tooling/Refactoring/Rename/RenamingAction.h]: I suspect you might get some push back on this, with the canonical declaration being returning an `Error` and taking a `SmallVector&`.

File: clang/lib/Tooling/Refactoring/Rename/RenamingAction.cpp [hunk 9]
```diff
@@ -275,5 +275,99 @@ std::unique_ptr<ASTConsumer> QualifiedRenamingAction::newASTConsumer() {
...
+      // Stop scanning at the end of the message send.
+      // Also account for spurious ']' in blocks or lambdas.
+      if (Kind == ObjCSymbolSelectorKind::MessageSend && !SquareCount &&
+          !BraceCount)
+        break;
+      if (SquareCount)
+        --SquareCount;
+    } else if (Tok.kind() == tok::l_square)
+      ++SquareCount;
+    else if (Tok.kind() == tok::l_paren)
```
Comment (bnbarham) [inline clang/lib/Tooling/Refactoring/Rename/RenamingAction.cpp]: I realize this is a copy paste from the fork implementation, but would be nice to add braces in these cases. All for none in simple one line returns but it's pretty weird here.

File: clang/lib/Tooling/Refactoring/Rename/RenamingAction.cpp [hunk 10]
```diff
@@ -275,5 +275,99 @@ std::unique_ptr<ASTConsumer> QualifiedRenamingAction::newASTConsumer() {
...
+      --BraceCount;
+    }
+    // Stop scanning at the end of the method's declaration.
+    if (Kind == ObjCSymbolSelectorKind::MethodDecl && NoScoping &&
+        (Tok.kind() == tok::semi || Tok.kind() == tok::minus ||
+         Tok.kind() == tok::plus))
+      break;
+  }
+  return llvm::make_error<llvm::StringError>(
+      "failed to find all selector pieces in the source code",
```
Comment (bnbarham) [inline clang/lib/Tooling/Refactoring/Rename/RenamingAction.cpp:368]: Any thoughts on attempting to be more lenient here? Do we *really* care if we can't match it exactly?

File: clang/lib/Tooling/Refactoring/SymbolName.cpp [hunk 11]
```diff
@@ -34,6 +39,21 @@ SymbolName::SymbolName(ArrayRef<StringRef> NamePieces) {
     this->NamePieces.push_back(Piece.str());
 }
 
+std::optional<std::string> SymbolName::getSinglePiece() const {
+  if (getNamePieces().size() == 1) {
+    return NamePieces.front();
+  } else {
+    return std::nullopt;
+  }
```
Comment (bnbarham) [inline clang/lib/Tooling/Refactoring/SymbolName.cpp]: No need for the `else`

File: clang-tools-extra/clangd/refactor/Rename.h [hunk 12]
```diff
@@ -66,13 +67,19 @@ struct RenameResult {
 llvm::Expected<RenameResult> rename(const RenameInputs &RInputs);
 
 /// Generates rename edits that replaces all given occurrences with the
-/// NewName.
+/// `NewName`.
+///
+/// `OldName` is and `Tokens` are used to to find the argument labels of
```
Comment (bnbarham) [inline clang-tools-extra/clangd/refactor/Rename.h]: Spurious `is`

File: clang-tools-extra/clangd/unittests/RenameTests.cpp [hunk 13]
```diff
@@ -1808,6 +1910,147 @@ TEST(CrossFileRenameTests, WithUpToDateIndex) {
   }
 }
 
+TEST(CrossFileRenameTests, ObjC) {
+  MockCompilationDatabase CDB;
+  CDB.ExtraClangFlags = {"-xobjective-c"};
+  // rename is runnning on all "^" points in FooH, and "[[]]" ranges are the
+  // expected rename occurrences.
```
Comment (bnbarham) [inline clang-tools-extra/clangd/unittests/RenameTests.cpp]: I don't see any `[[]]`

File: clang/include/clang/Tooling/Syntax/Tokens.h [hunk 14]
```diff
@@ -145,6 +145,20 @@ class Token {
 /// For debugging purposes. Equivalent to a call to Token::str().
 llvm::raw_ostream &operator<<(llvm::raw_ostream &OS, const Token &T);
 
+/// A list of tokens as lexed from the input file, without expanding
+/// preprocessor macros.
+class UnexpandedTokenBuffer {
+  std::string Code;
```
Comment (bnbarham) [inline clang/include/clang/Tooling/Syntax/Tokens.h]: Do we really need to take a copy of `Code` here?

File: clang/lib/Tooling/Syntax/Tokens.cpp [hunk 15]
```diff
@@ -225,6 +225,17 @@ llvm::StringRef FileRange::text(const SourceManager &SM) const {
   return Text.substr(Begin, length());
 }
 
+UnexpandedTokenBuffer::UnexpandedTokenBuffer(StringRef Code,
+                                             const LangOptions &LangOpts) {
+  // InMemoryFileAdapter crashes unless the buffer is null terminated, so ensure
+  // the string is null-terminated.
```
Comment (bnbarham) [inline clang/lib/Tooling/Syntax/Tokens.cpp]: `InMemoryFileAdapter` doesn't exist any more and I believe `MemoryBuffer` does handle not having a null terminator. We know the size the whole way through so seems weird that we wouldn't.

File: clang-tools-extra/clangd/refactor/Rename.cpp [hunk 16]
```diff
@@ -811,7 +811,20 @@ llvm::Expected<RenameResult> rename(const RenameInputs &RInputs) {
   if (!Name)
     return makeError(ReasonToReject::UnsupportedSymbol);
   SymbolName OldSymbolName(Name);
-  SymbolName NewSymbolName(RInputs.NewName, AST.getLangOpts());
+  SymbolName NewSymbolName(ArrayRef<StringRef>{});
```
Comment (bnbarham) [inline clang-tools-extra/clangd/refactor/Rename.cpp]: Can't just do `{}`? Wouldn't hurt to have a no-arg constructor IMO

File: clang-tools-extra/clangd/refactor/Rename.cpp [hunk 17]
```diff
@@ -811,7 +811,20 @@ llvm::Expected<RenameResult> rename(const RenameInputs &RInputs) {
...
+  if (std::optional<StringRef> NewName = RInputs.NewName) {
+    NewSymbolName = SymbolName(*NewName, AST.getLangOpts());
+  } else {
+    // If no new name is given, we are perfoming a pseudo rename for the
+    // prepareRename request to check if the rename is possible. Construct a
+    // new symbol name that has as many name pieces as the old name and is thus
+    // a valid new name.
+    std::vector<std::string> NewNamePieces;
+    for (StringRef Piece : OldSymbolName.getNamePieces()) {
+      NewNamePieces.push_back(Piece.str() + "__clangd_rename_placeholder");
```
Comment (bnbarham) [inline clang-tools-extra/clangd/refactor/Rename.cpp]: Isn't `RenameInputs::NewName` to handle this case? If not, what about just adding the placeholder name to the very first piece?

File: clang-tools-extra/clangd/unittests/RenameTests.cpp [hunk 18]
```diff
@@ -1507,6 +1508,39 @@ TEST(RenameTest, PrepareRename) {
...
+  trace::TestTracer Tracer;
+  MockCompilationDatabase CDB;
+  CDB.ExtraClangFlags = {"-xobjective-c"};
+  ClangdServer Server(CDB, FS, ServerOpts);
+  runAddDocument(Server, Path, Input.code());
+
+  for (Position Point : Input.points()) {
+    auto Results = runPrepareRename(Server, Path, Point,
+                                    /*NewName=*/std::nullopt, {});
+    // Verify that for multi-file rename, we only return main-file occurrences.
```
Comment (bnbarham) [inline clang-tools-extra/clangd/unittests/RenameTests.cpp]: Aren't we only checking a single file here?

**Review by bnbarham — COMMENTED** [review]
Nice!

File: clang/lib/Tooling/Refactor/RenameIndexedFile.cpp (hunk 1 above)
Comment (ahoppen) [reply clang/lib/Tooling/Refactor/RenameIndexedFile.cpp:268]: I want to upstream `SymbolName` but `containsEmptyPiece` is only used in the `Refactor` module which won’t be upstreamed. And I didn’t want to add a function to `SymbolName` that doesn’t have any use upstream.

File: clang/include/clang/Tooling/Refactoring/Rename/SymbolName.h (hunk 3 above)
Comment (ahoppen) [reply clang/include/clang/Tooling/Refactoring/Rename/SymbolName.h:33]: The upstreaming commit will not resemble any apple/llvm-project commit anyway because in our fork, we need to adjust the uses of the Apple-only `OldSymbolName` at the same time. I’ll just untangle all of this while upstreaming. Should be easier than trying to find the boundaries right now.

File: clang-tools-extra/clangd/refactor/Rename.cpp (hunk 5 above)
Comment (ahoppen) [reply clang-tools-extra/clangd/refactor/Rename.cpp:585]: We can’t early return because there’s the surrounding loop

```cpp
for (SourceLocation Loc : findOccurrencesWithinFile(AST, RenameDecl)) {
```

We could `continue` in the `if` block. But I don’t really like that.

File: clang-tools-extra/clangd/unittests/RenameTests.cpp (hunk 7 above)
Comment (ahoppen) [reply clang-tools-extra/clangd/unittests/RenameTests.cpp:1085]: That’s the whole purpose of the PR. We used to error because we couldn’t rename Obj-C symbols and now we can. So there’s no error anymore.

File: clang/lib/Tooling/Refactoring/Rename/RenamingAction.cpp (hunk 10 above)
Comment (ahoppen) [reply clang/lib/Tooling/Refactoring/Rename/RenamingAction.cpp:368]: I think if we can’t find all the selector pieces than something is going wrong and that location doesn’t actually contain a reference to the selector we are looking for. Instead of e.g. partially renaming it, which will definitely result in invalid code, I think it’s better to just leave it.

File: clang/include/clang/Tooling/Syntax/Tokens.h (hunk 14 above)
Comment (ahoppen) [reply clang/include/clang/Tooling/Syntax/Tokens.h]: Whoops, copy paste error.

File: clang-tools-extra/clangd/unittests/RenameTests.cpp (hunk 7 above)
Comment (bnbarham) [reply clang-tools-extra/clangd/unittests/RenameTests.cpp:1085]: Must have been tired when I was reviewing this, I think I saw "fo^o" and thought "huh, that should still error". But that's just the location 😅

File: clang-tools-extra/clangd/refactor/Rename.cpp (hunk 16 above)
Comment (ahoppen) [reply clang-tools-extra/clangd/refactor/Rename.cpp]: Added a no-arg constructor.

File: clang-tools-extra/clangd/unittests/RenameTests.cpp (hunk 18 above)
Comment (ahoppen) [reply clang-tools-extra/clangd/unittests/RenameTests.cpp]: Just another copy-paste error.

File: clang-tools-extra/clangd/unittests/RenameTests.cpp (hunk 13 above)
Comment (ahoppen) [reply clang-tools-extra/clangd/unittests/RenameTests.cpp]: Copy-paste error.

**Comment by ahoppen** [comment]
@swift-ci Please test

File: clang/lib/Tooling/Refactoring/Rename/RenamingAction.cpp (hunk 10 above)
Comment (bnbarham) [reply clang/lib/Tooling/Refactoring/Rename/RenamingAction.cpp:368]: As long as that only skips this reference and doesn't fail the whole rename, that sounds fine to me 👍

**Review by bnbarham — APPROVED** [review]
Thanks!

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows
